pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,237 @@
+		defer sp.Finish()
+
+		resp, err := d.doRequest(localCtx, am, &httpgrpc.HTTPRequest{
+			Method:  r.Method,
+			Url:     r.RequestURI,
+			Body:    body,
+			Headers: grpcHeaders,
+		})
+		if err != nil {
+			return err
+		}
+
+		if resp.Code/100 != 2 {
+			return httpgrpc.ErrorFromHTTPResponse(resp)
+		}
+
+		respMtx.Lock()
+		if firstSuccessfulResponse == nil {
+			firstSuccessfulResponse = resp
+		}
+		respMtx.Unlock()
+
+		return nil
+	}, func() {})
+
+	if err != nil {
+		respondFromError(err, w, logger)
+	}
+
+	respMtx.Lock() // Another request might be ongoing after quorum.
+	resp := firstSuccessfulResponse
+	respMtx.Unlock()
+
+	if resp != nil {
+		respondFromHTTPGRPCResponse(w, resp)
+	} else {
+		// This should not happen.
+		level.Error(logger).Log("msg", "distributor did not receive response from alertmanager though no errors")
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, RingOp, nil, nil, nil)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get replication set from the ring", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	req, err := server.HTTPRequest(r)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get grpc request from http request", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(r.Context(), d.cfg.AlertmanagerClient.RemoteTimeout)
+	defer cancel()
+
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "DistributeRequest.doRead")
+	defer sp.Finish()
+	// Until we have a mechanism to combine the results from multiple alertmanagers,
+	// we forward the request to only only of the alertmanagers.
+	amDesc := replicationSet.Ingesters[rand.Intn(len(replicationSet.Ingesters))]
+	resp, err := d.doRequest(ctx, amDesc, req)
+	if err != nil {
+		respondFromError(err, w, logger)
+		return
+	}
+
+	respondFromHTTPGRPCResponse(w, resp)
+}
+
+func respondFromError(err error, w http.ResponseWriter, logger log.Logger) {
+	httpResp, ok := httpgrpc.HTTPResponseFromError(errors.Cause(err))
+	if !ok {
+		level.Error(logger).Log("msg", "failed to process the request to the alertmanager", "err", err)
+		http.Error(w, "Failed to process the request to the alertmanager", http.StatusInternalServerError)
+		return
+	}
+	respondFromHTTPGRPCResponse(w, httpResp)
+}
+
+func respondFromHTTPGRPCResponse(w http.ResponseWriter, httpResp *httpgrpc.HTTPResponse) {
+	for _, h := range httpResp.Headers {
+		for _, v := range h.Values {
+			w.Header().Add(h.Key, v)
+		}
+	}
+	w.WriteHeader(int(httpResp.Code))
+	w.Write(httpResp.Body) //nolint
+}
+
+func (d *Distributor) doRequest(ctx context.Context, am ring.InstanceDesc, req *httpgrpc.HTTPRequest) (*httpgrpc.HTTPResponse, error) {
```

Comment:
[nit] Since you need to always apply the `RemoteTimeout` and since `doRequest()` takes `Distributor` as receiver (and you have access to the distributor config) you could apply the `RemoteTimeout` here so that we do it only once and for all.